nncf/quantization/quantizer_propagation.py

```diff
@@ -798,11 +1174,28 @@ def propagation_step(self, curr_prop_quantizer: PropagatingQuantizer,
            some insertion point."""
         # TODO: full-fledged discrete finite automata approach? Switch to traversing a graph
         # consisting of insertion points only, with reversed edges holding associated operator data?
+
+        #pylint:disable=too-many-branches
         curr_node_key = curr_prop_quantizer.current_location_node_key
         curr_node = quant_prop_graph.nodes[curr_prop_quantizer.current_location_node_key]
         curr_node_type = curr_node[QuantizerPropagationStateGraph.NODE_TYPE_NODE_ATTR]
         assert curr_node_type == QuantizerPropagationStateGraphNodeType.INSERTION_POINT
 
+        if curr_prop_quantizer in self._quantizers_waiting_for_branch_merge:
+            branching_node_key = self._quantizers_waiting_for_branch_merge.get_blocking_node(curr_prop_quantizer)
+            dom_pqs = quant_prop_graph.get_propagating_quantizers_immediately_dominated_by_node(branching_node_key)
+            active_dom_pqs = set(
+                filter(lambda x: x in self._active_propagating_quantizers_queue or x is curr_prop_quantizer, dom_pqs))
+            waiting_pqs = self._quantizers_waiting_for_branch_merge.get_waiting_quantizers_for_branching_node_key(
+                branching_node_key)
+            if waiting_pqs == active_dom_pqs:
+                self._active_propagating_quantizers_queue.append(curr_prop_quantizer)
+                self._handle_quantizer_merge(waiting_pqs, quant_prop_graph, branching_node_key)
+            else:
+                # Not all of the dominated quantizers have reached the blocking node yet
+                self._active_propagating_quantizers_queue.appendleft(curr_prop_quantizer)
+            return quant_prop_graph
```

Comment:
2) The new propagation algorithm will make all quantizers that are trying to pass up through a downward-branching node wait until the quantizers on the neighbouring branches reach that node as well. Only then will the merge process occur, which may result in the new merge quantizer being added above the branching node to propagate further, and in the existing branch quantizers being discarded. For the `PropagationStrategy.AGGRESSIVE`, it is also possible that neither of the branch quantizers is discarded, but the merge quantizer is still created - for instance, if one of the branches will potentially require requantization, in case the user/specialized algorithm chooses a corresponding configuration.

![aggressive_merge](https://user-images.githubusercontent.com/31695470/100772821-24b71a00-3411-11eb-9902-24a4d4fd4795.png)

Here, the merge quantizer will have a single 8-bit configuration associated since this is the only option that is supported by all the downstream branches. If, at the final quantization config selection stage of the setup, some of the branches also end up having 8-bit quantization chosen for them, the corresponding branch quantizers will be discarded since the global quantizer already has 8-bit quantization. The user is still, however, presented with a possibility to choose non-8-bit quantization for all the branches, in which case requantization will occur on the branches.